lib/Dialect/ESI/runtime/cpp/include/esi/Types.h

```diff
@@ -33,6 +39,26 @@ class Type {
   ID getID() const { return id; }
   virtual std::ptrdiff_t getBitWidth() const { return -1; }
 
+  /// Serialize an object to MessageData. The object should be passed as a
+  /// std::any to provide type erasure. Returns a MessageData containing the
+  /// serialized representation.
+  virtual MessageData serialize(const std::any &obj) const {
+    throw std::runtime_error("Serialization not implemented for type " + id);
+  }
+
+  /// Deserialize from a span of bytes to an object. Returns the deserialized
+  /// object as a std::any and a span to the remaining bytes.
+  virtual std::pair<std::any, std::span<const uint8_t>>
+  deserialize(std::span<const uint8_t> data) const {
+    throw std::runtime_error("Deserialization not implemented for type " + id);
+  }
+
+  /// Check if a std::any object is valid for this type. Throws
+  /// std::runtime_error if the object is not valid.
+  virtual void isValid(const std::any &obj) const {
```

Comment:
One of the reasons the Python API is the way it is is because it doesn't make sense to have a method called `isValid` throw an exception rather than return bool. If we're going to go this way, rename it to `checkValid` or `ensureValid`. We can then have an `isValid` with a try-catch.

Though I prefer it the other way. I don't like relying on exceptions to do something perfectly valid. "Exceptions should be for exceptional circumstances" -Some book I once read.

Reply:
> Though I prefer it the other way. I don't like relying on exceptions to do something perfectly valid. "Exceptions should be for exceptional circumstances" -Some book I once read.

I used to be of the same opinion, but have lately come to, errr, embrace exceptions more. I think they're a perfectly valid, and extremely useful, way to get another form of control flow without going overboard in C++ boilerplate.